### PR TITLE
Umcs 284 CIT/MIT Add recurrance type to Java SDK

### PR DESCRIPTION
**Pull request by DerAusHeidelberg**

### CODE REVIEW

**Review by mghazzouz — CHANGES_REQUESTED** [review]
@DerAusHeidelberg  First remark, please check failing checks in "Codacy Static Code Analysis"

https://app.codacy.com/gh/unzerdev/java-sdk/pullRequest?prid=7995522

**Review by Ryouzanpaku — COMMENTED** [review]
looking fine, but the version seem to not fit.

File: src/main/java/com/unzer/payment/Unzer.java [hunk 1]
```diff
@@ -1039,14 +1111,19 @@ public Linkpay linkpay (Linkpay linkpay) throws HttpCommunicationException {
 	public Recurring recurring(String typeId, String customerId, String metadataId, URL returnUrl) throws HttpCommunicationException {
 		return paymentService.recurring(getRecurring(typeId, customerId, metadataId, returnUrl));
 	}
+
+    public Recurring recurring(String typeId, String customerId, String metadataId, URL returnUrl, AdditionalTransactionData additionalTransactionData) throws HttpCommunicationException {
+        return paymentService.recurring(getRecurring(typeId, customerId, metadataId, returnUrl, additionalTransactionData));
```
[the remark this answers is not in the document]
Comment (DerAusHeidelberg) [reply src/main/java/com/unzer/payment/Unzer.java:1116]: This is just visually here on Github. In code this is fine. I also format the Classes before Saving.

**Review by Ryouzanpaku — APPROVED** [review]
Lgtm.

File: src/test/java/com/unzer/payment/business/paymenttypes/CardTest.java [hunk 2]
```diff
@@ -216,4 +228,255 @@ public void testCardMailInvalidValue() throws HttpCommunicationException {
...
+
+    @Test
+    public void testCardRegisterRecurring() throws HttpCommunicationException, MalformedURLException {
+        Card card = new Card(VISA_3DS_ENABLED_CARD_NUMBER, "03/99");
+        card.setCvc("123");
+        card = getUnzer().createPaymentType(card);
+
+        Recurring recurring = getUnzer().recurring(card.getId(), null, null, new URL("https://www.meinShop.de"));
+    }
+
```
Comment (mghazzouz) [inline src/test/java/com/unzer/payment/business/paymenttypes/CardTest.java:240]: JUnit tests should include assert() or fail()

see.. https://app.codacy.com/gh/unzerdev/java-sdk/pullRequest?prid=7995522

**Review by mghazzouz — CHANGES_REQUESTED** [review]

File: src/main/java/com/unzer/payment/models/CardTransactionData.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+@JsonTypeName("card")
+@JsonTypeInfo(include = JsonTypeInfo.As.WRAPPER_OBJECT, use = JsonTypeInfo.Id.NAME)
+public class CardTransactionData {
+
+    public CardTransactionData(RecurrenceType recurrenceType) {
+        this.recurrenceType = recurrenceType;
+    }
+
+    RecurrenceType recurrenceType;
```
Comment (mghazzouz) [inline src/main/java/com/unzer/payment/models/CardTransactionData.java:15]: recurrenceType to be declared at the top of the class, before constructor

see https://app.codacy.com/gh/unzerdev/java-sdk/pullRequest?prid=7995522

**Review by mghazzouz — CHANGES_REQUESTED** [review]

File: src/main/java/com/unzer/payment/models/CardTransactionData.java (hunk 3 above)
Comment (DerAusHeidelberg) [reply src/main/java/com/unzer/payment/models/CardTransactionData.java]: done

File: src/test/java/com/unzer/payment/business/paymenttypes/CardTest.java (hunk 2 above)
Comment (DerAusHeidelberg) [reply src/test/java/com/unzer/payment/business/paymenttypes/CardTest.java]: done

File: src/main/java/com/unzer/payment/ApplePaySession.java [hunk 4]
```diff
@@ -1,5 +1,25 @@
 package com.unzer.payment;
 
+/*-
+ * #%L
+ * Unzer Java SDK
+ * %%
+ * Copyright (C) 2020 - 2021 Unzer E-Com GmbH
```
Comment (sixer1182) [inline src/main/java/com/unzer/payment/ApplePaySession.java:7]:
```suggestion
 * Copyright (C) 2020 - today Unzer E-Com GmbH
```

File: src/main/java/com/unzer/payment/AbstractPayment.java [hunk 5]
```diff
@@ -1,5 +1,25 @@
 package com.unzer.payment;
 
+/*-
+ * #%L
+ * Unzer Java SDK
+ * %%
+ * Copyright (C) 2020 - 2021 Unzer E-Com GmbH
```
Comment (sixer1182) [inline src/main/java/com/unzer/payment/AbstractPayment.java:7]:
```suggestion
 * Copyright (C) 2020 - today Unzer E-Com GmbH
```

**Review by mghazzouz — DISMISSED** [review]

File: src/main/java/com/unzer/payment/communication/JsonFieldIgnore.java [hunk 6]
```diff
@@ -1,5 +1,25 @@
 package com.unzer.payment.communication;
 
+/*-
+ * #%L
+ * Unzer Java SDK
+ * %%
+ * Copyright (C) 2020 - 2021 Unzer E-Com GmbH
```
Comment (sixer1182) [inline src/main/java/com/unzer/payment/communication/JsonFieldIgnore.java:7]:
```suggestion
 * Copyright (C) 2020 - today Unzer E-Com GmbH
```

File: src/main/java/com/unzer/payment/communication/JsonFieldIgnoreStragegy.java [hunk 7]
```diff
@@ -1,5 +1,25 @@
 package com.unzer.payment.communication;
 
+/*-
+ * #%L
+ * Unzer Java SDK
+ * %%
+ * Copyright (C) 2020 - 2021 Unzer E-Com GmbH
```
Comment (sixer1182) [inline src/main/java/com/unzer/payment/communication/JsonFieldIgnoreStragegy.java:7]:
```suggestion
 * Copyright (C) 2020 - today Unzer E-Com GmbH
```

File: src/main/java/com/unzer/payment/communication/json/JsonInstallmentSecuredRatePlanList.java [hunk 8]
```diff
@@ -1,5 +1,25 @@
 package com.unzer.payment.communication.json;
 
+/*-
+ * #%L
+ * Unzer Java SDK
+ * %%
+ * Copyright (C) 2020 - 2021 Unzer E-Com GmbH
```
Comment (sixer1182) [inline src/main/java/com/unzer/payment/communication/json/JsonInstallmentSecuredRatePlanList.java:7]:
```suggestion
 * Copyright (C) 2020 - today Unzer E-Com GmbH
```

File: src/main/java/com/unzer/payment/enums/RecurrenceType.java [hunk 9]
```diff
@@ -0,0 +1,32 @@
+package com.unzer.payment.enums;
+
+/*-
+ * #%L
+ * Unzer Java SDK
+ * %%
+ * Copyright (C) 2020 - 2021 Unzer E-Com GmbH
```
Comment (sixer1182) [inline src/main/java/com/unzer/payment/enums/RecurrenceType.java:7]:
```suggestion
 * Copyright (C) 2020 - today Unzer E-Com GmbH
```

File: src/main/java/com/unzer/payment/marketplace/MarketplaceAuthorization.java [hunk 10]
```diff
@@ -1,5 +1,25 @@
 package com.unzer.payment.marketplace;
 
+/*-
+ * #%L
+ * Unzer Java SDK
+ * %%
+ * Copyright (C) 2020 - 2021 Unzer E-Com GmbH
```
Comment (sixer1182) [inline src/main/java/com/unzer/payment/marketplace/MarketplaceAuthorization.java:7]:
```suggestion
 * Copyright (C) 2020 - today Unzer E-Com GmbH
```

File: src/main/java/com/unzer/payment/marketplace/MarketplaceCancel.java [hunk 11]
```diff
@@ -1,5 +1,25 @@
 package com.unzer.payment.marketplace;
 
+/*-
+ * #%L
+ * Unzer Java SDK
+ * %%
+ * Copyright (C) 2020 - 2021 Unzer E-Com GmbH
```
Comment (sixer1182) [inline src/main/java/com/unzer/payment/marketplace/MarketplaceCancel.java:7]:
```suggestion
 * Copyright (C) 2020 - today Unzer E-Com GmbH
```

File: src/main/java/com/unzer/payment/marketplace/MarketplaceCancelBasket.java [hunk 12]
```diff
@@ -1,5 +1,25 @@
 package com.unzer.payment.marketplace;
 
+/*-
+ * #%L
+ * Unzer Java SDK
+ * %%
+ * Copyright (C) 2020 - 2021 Unzer E-Com GmbH
```
Comment (sixer1182) [inline src/main/java/com/unzer/payment/marketplace/MarketplaceCancelBasket.java:7]:
```suggestion
 * Copyright (C) 2020 - today Unzer E-Com GmbH
```

**Review by sixer1182 — CHANGES_REQUESTED** [review]
LGTM - Please update all copyrights, I did not mark them all.

File: src/main/java/com/unzer/payment/ApplePaySession.java (hunk 4 above)
Comment (DerAusHeidelberg) [reply src/main/java/com/unzer/payment/ApplePaySession.java]: changed

File: src/main/java/com/unzer/payment/AbstractPayment.java (hunk 5 above)
Comment (DerAusHeidelberg) [reply src/main/java/com/unzer/payment/AbstractPayment.java]: changed

File: src/main/java/com/unzer/payment/communication/JsonFieldIgnore.java (hunk 6 above)
Comment (DerAusHeidelberg) [reply src/main/java/com/unzer/payment/communication/JsonFieldIgnore.java]: changed

File: src/main/java/com/unzer/payment/communication/JsonFieldIgnoreStragegy.java (hunk 7 above)
Comment (DerAusHeidelberg) [reply src/main/java/com/unzer/payment/communication/JsonFieldIgnoreStragegy.java]: changed

File: src/main/java/com/unzer/payment/communication/json/JsonInstallmentSecuredRatePlanList.java (hunk 8 above)
Comment (DerAusHeidelberg) [reply src/main/java/com/unzer/payment/communication/json/JsonInstallmentSecuredRatePlanList.java]: changed

File: src/main/java/com/unzer/payment/enums/RecurrenceType.java (hunk 9 above)
Comment (DerAusHeidelberg) [reply src/main/java/com/unzer/payment/enums/RecurrenceType.java]: changed

File: src/main/java/com/unzer/payment/marketplace/MarketplaceAuthorization.java (hunk 10 above)
Comment (DerAusHeidelberg) [reply src/main/java/com/unzer/payment/marketplace/MarketplaceAuthorization.java]: changed

File: src/main/java/com/unzer/payment/marketplace/MarketplaceCancel.java (hunk 11 above)
Comment (DerAusHeidelberg) [reply src/main/java/com/unzer/payment/marketplace/MarketplaceCancel.java]: changed

File: src/main/java/com/unzer/payment/marketplace/MarketplaceCancelBasket.java (hunk 12 above)
Comment (DerAusHeidelberg) [reply src/main/java/com/unzer/payment/marketplace/MarketplaceCancelBasket.java]: changed

File: src/main/java/com/unzer/payment/service/UrlUtil.java [hunk 13]
```diff
@@ -21,11 +21,12 @@
  * %%
  * Copyright (C) 2020 Unzer E-Com GmbH
```
Comment (sixer1182) [inline src/main/java/com/unzer/payment/service/UrlUtil.java:22]:
```suggestion
 * Copyright (C) 2020 - today Unzer E-Com GmbH
```

**Review by sixer1182 — APPROVED** [review]
Looks good 💯

**Review by Ryouzanpaku — APPROVED** [review]
LGTM.

**Pull request merged**